File: pdl/infra/php/src/Io/Pdl/Infrastructure/Data/Row.php
<?php

namespace Io\Pdl\Infrastructure\Data;

use RuntimeException;

class Row
{
    private array $propertyToColumn;
    private array $columnToProperty;
    private ?DBStore $store;

    public function __construct(
        private string $table,
        private string $primaryKey,
        array $columnMap,
        ?DBStore $store = null
    ) {
        $this->propertyToColumn = $columnMap;
        $this->columnToProperty = array_change_key_case(array_flip($columnMap));
        $this->store = $store;
    }

    public function table(): string
    {
        return $this->table;
    }

    public function primaryKey(): string
    {
        return $this->primaryKey;
    }

    public function store(): ?DBStore
    {
        return $this->store;
    }

    public function setStore(?DBStore $store): void
    {
        $this->store = $store;
    }

    public function collectValues(): array
    {
        $values = [];
        foreach ($this->propertyToColumn as $property => $column) {
            if (!property_exists($this, $property)) {
                continue;
            }
            $value = $this->{$property};
            $values[$column] = $value;
        }
        return $values;
    }

    public function applyValues(array $values): void
    {
        foreach ($values as $column => $value) {
            $normalized = strtolower($column);
            $property = $this->columnToProperty[$normalized] ?? $this->derivePropertyName($column);
            if ($property === null || !property_exists($this, $property)) {
                continue;
            }
            $this->{$property} = $value;
        }
    }

    private function derivePropertyName(string $column): ?string
    {
        $normalized = strtolower($column);
        if (isset($this->columnToProperty[$normalized])) {
            return $this->columnToProperty[$normalized];
        }
        $segments = explode('_', $normalized);
        $segments = array_map(static fn(string $part): string => ucfirst($part), $segments);
        $property = lcfirst(implode('', $segments));
        return property_exists($this, $property) ? $property : null;
    }

    public function resolveStore(): DBStore
    {
        return StoreRegistry::resolve($this->store);
    }
}

File: pdl/infra/csharp/Io/Pdl/Infrastructure/Data/Row.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;

namespace Io.Pdl.Infrastructure.Data;

public abstract class Row
{
    private readonly IReadOnlyDictionary<string, string> _propertyToColumn;
    private readonly IReadOnlyDictionary<string, string> _columnToProperty;
    private IDbStore? _store;

    protected Row(string table, string primaryKey, IReadOnlyDictionary<string, string> columnMap, IDbStore? store = null)
    {
        Table = table;
        PrimaryKey = primaryKey;
        _propertyToColumn = columnMap;
        _columnToProperty = columnMap.ToDictionary(pair => pair.Value.ToLowerInvariant(), pair => pair.Key);
        _store = store;
    }

    public string Table { get; }

    public string PrimaryKey { get; }

    public IDbStore? Store => _store;

    public void SetStore(IDbStore? store)
    {
        _store = store;
    }

    public IDictionary<string, object?> CollectValues()
    {
        var values = new Dictionary<string, object?>();
        foreach (var (property, column) in _propertyToColumn)
        {
            var propertyInfo = GetType().GetProperty(property);
            if (propertyInfo == null)
            {
                continue;
            }

            values[column] = propertyInfo.GetValue(this);
        }

        return values;
    }

    public void ApplyValues(IDictionary<string, object?> values)
    {
        foreach (var (column, value) in values)
        {
            var key = column.ToLower(CultureInfo.InvariantCulture);
            if (!_columnToProperty.TryGetValue(key, out var property))
            {
                property = DerivePropertyName(column);
                if (property == null)
                {
                    continue;
                }
            }

            var propertyInfo = GetType().GetProperty(property);
            propertyInfo?.SetValue(this, value);
        }
    }

    public IDbStore ResolveStore()
    {
        return StoreRegistry.Resolve(_store);
    }

    private string? DerivePropertyName(string column)
    {
        var segments = column.Split('_');
        var property = string.Concat(segments.Select(static segment => char.ToUpperInvariant(segment[0]) + segment[1..].ToLowerInvariant()));
        if (property.Length == 0)
        {
            return null;
        }

        property = char.ToLowerInvariant(property[0]) + property[1..];
        return GetType().GetProperty(property) != null ? property : null;
    }
}
